third_party/move/move-model/src/ty.rs

```diff
@@ -2130,77 +2658,175 @@ impl TypeUnificationError {
-                        WideningOrder::Join | WideningOrder::LeftToRight => (options_str, type_str),
-                        WideningOrder::RightToLeft => (type_str, options_str),
-                    };
-                    error_context.type_mismatch_str(display_context, actual, expected)
-                },
-                Constraint::SomeReference(ty) => {
-                    error_context.expected_reference(display_context, ty)
-                },
-                Constraint::SomeStruct(field_map) => {
-                    Self::message_for_struct(unification_context, display_context, field_map, ty)
-                },
-                Constraint::SomeReceiverFunction(name, ..) => {
+                        .to_string(),
+                    vec![],
+                    vec![],
+                )
+            },
+            TypeUnificationError::MutabilityMismatch(actual, expected) => (
+                error_context.mutability_mismatch(*actual, *expected),
+                vec![],
+                vec![],
+            ),
+            TypeUnificationError::MissingAbilities(ty, missing, ctx_opt) => {
+                let (note, hints, labels) = ctx_opt
+                    .as_ref()
+                    .map(|ctx| ctx.describe(display_context))
+                    .unwrap_or_default();
+                (
                     format!(
-                        "undeclared receiver function `{}` for type `{}`",
-                        name.display(display_context.env.symbol_pool()),
-                        ty.display(display_context)
-                    )
-                },
-                Constraint::HasAbility(ability) => format!(
-                    "type `{}` does not have expected ability `{}`",
-                    ty.display(display_context),
-                    ability
-                ),
-                Constraint::WithDefault(_) => unreachable!("default constraint in error message"),
+                        "type `{}` is missing required {} `{}`{}",
+                        ty.display(display_context),
+                        pluralize("ability", missing.iter().count()),
+                        missing,
+                        if !note.is_empty() {
+                            format!(" ({})", note)
+                        } else {
+                            "".to_string()
+                        }
+                    ),
+                    hints,
+                    labels,
+                )
+            },
+            TypeUnificationError::ConstraintUnsatisfied(_, ty, order, constr, ctx_opt) => {
+                let item_name = || match ctx_opt {
+                    Some(ConstraintContext {
+                        origin: ConstraintOrigin::Field(_),
+                        ..
+                    }) => "as a field type",
+                    Some(ConstraintContext {
+                        origin: ConstraintOrigin::Local(_),
+                        ..
+                    }) => "as a local type",
```

Comment:
`"as a local variable type"` would be better, since Move developers wouldn't know what a "local type" is.

Reply:
Done